### PR TITLE
multi GSTIN Using journal

### PR DESCRIPTION
**Pull request by jva-odoo**
task https://www.odoo.com/web#id=1917619&action=333&active_id=1691&model=project.task&view_type=form&menu_id=4720

pad https://pad.odoo.com/p/r.9da01f115c928cde4b8eb85243999c84

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_in/models/chart_template.py [hunk 1]
```diff
@@ -11,12 +11,13 @@ def _prepare_all_journals(self, acc_template_ref, company, journals_dict=None):
         res = super(AccountChartTemplate, self)._prepare_all_journals(acc_template_ref, company, journals_dict=journals_dict)
         if self == self.env.ref('l10n_in.indian_chart_template_standard'):
             for journal in res:
+                journal['l10n_in_gstin_partner_id'] = company.partner_id.id
```
Comment (hmo-odoo) [inline addons/l10n_in/models/chart_template.py]: set value of l10n_in_gstin_partner_id for only sales and purchase journal

File: addons/l10n_in/models/company.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    l10n_in_gstin_partner_ids = fields.One2many('res.partner', 'l10n_in_gstin_company_id', string="Multi GSTIN")
+
+    @api.model
+    def create(self, vals):
+        # company's partner will now act as it's gstin partner
+        partner = self.env['res.partner']
+        if vals.get('partner_id'):
+            partner = partner.browse(vals['partner_id'])
+        company = super(ResCompany, self).create(vals)
+        (partner or company.partner_id).sudo().write({'l10n_in_gstin_company_id': company})
```
Comment (hmo-odoo) [inline addons/l10n_in/models/company.py]: This for only Indian Company. so check company country is India or not

File: addons/l10n_in/models/res_config_settings.py [hunk 3]
```diff
@@ -9,6 +9,23 @@ class ResConfigSettings(models.TransientModel):
...
     show_module_l10n_in = fields.Boolean(compute='_compute_show_module_l10n_in')
     group_l10n_in_reseller = fields.Boolean(implied_group='l10n_in.group_l10n_in_reseller', string="Manage Reseller(E-Commerce)")
+    group_l10n_in_multi_gstin = fields.Boolean(string="Multi GSTIN", implied_group='l10n_in.group_l10n_in_multi_gstin')
+
+    def set_values(self):
+        super().set_values()
+        #set company it self as unit in exsiting records
+        if self.group_l10n_in_multi_gstin:
+            for company in self.env.user.company_ids:
+                company.partner_id.sudo().write({'l10n_in_gstin_company_id': company})
```
Comment (hmo-odoo) [inline addons/l10n_in/models/res_config_settings.py]: no need to set value of l10n_in_gstin_company_id. so remove set_values function here

File: addons/l10n_in/models/account_invoice.py [hunk 4]
```diff
@@ -28,6 +28,7 @@ def _compute_amount_total_words(self):
     l10n_in_shipping_port_code_id = fields.Many2one('l10n_in.port.code', 'Shipping port code', states={'draft': [('readonly', False)]})
     l10n_in_reseller_partner_id = fields.Many2one('res.partner', 'Reseller', domain=[('vat', '!=', False)], help="Only Registered Reseller", readonly=True, states={'draft': [('readonly', False)]})
     l10n_in_partner_vat = fields.Char(related="partner_id.vat", readonly=True)
+    l10n_in_gstin_partner_id = fields.Many2one(related="journal_id.l10n_in_gstin_partner_id")
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py]: group by a journal because of one GSTIN have at least one Journal

File: addons/l10n_in/models/company.py (hunk 2 above)
Comment (jva-odoo) [reply addons/l10n_in/models/company.py]: ok then also need to do in write

File: addons/l10n_in/models/res_config_settings.py (hunk 3 above)
Comment (jva-odoo) [reply addons/l10n_in/models/res_config_settings.py]: ok 👍

File: addons/l10n_in/models/chart_template.py (hunk 1 above)
Comment (jva-odoo) [reply addons/l10n_in/models/chart_template.py]: Done

File: addons/l10n_in/models/account.py [hunk 5]
```diff
@@ -10,7 +10,17 @@ class AccountJournal(models.Model):
...
+
+    @api.depends('name', 'state')
+    def name_get(self):
+        result = []
+        for journal in self:
+            name = journal.name
+            if journal.l10n_in_gstin_partner_id and journal.l10n_in_gstin_partner_id.vat:
+                name = "%s - %s" % (journal.name, journal.l10n_in_gstin_partner_id.vat)
+            result.append((journal.id, name))
+        return result
```
Comment (jco-odoo) [inline addons/l10n_in/models/account.py]: Better to have an onchange

Comment (jco-odoo) [reply addons/l10n_in/models/account.py]: India specific without depends

File: addons/l10n_in/models/res_partner.py [hunk 6]
```diff
@@ -10,6 +10,7 @@ class ResPartner(models.Model):
 
     # Use in view attrs. Need to required state_id if Country is India.
     l10n_in_country_code = fields.Char(related="country_id.code", string="Country code")
+    l10n_in_gstin_company_id = fields.Many2one('res.company', string="Company of Current GSTIN", ondelete='cascade')
```
Comment (jco-odoo) [inline addons/l10n_in/models/res_partner.py]: we can probably remove

File: addons/l10n_in/report/account_invoice_report.py [hunk 7]
```diff
@@ -259,14 +257,12 @@ def _from(self):
 
     def _where(self):
         return """
-                WHERE am.state = 'posted'
-                    AND tag_rep_ln.account_tax_report_line_id in (SELECT res_id FROM ir_model_data WHERE module='l10n_in' AND name in ('tax_report_line_igst','tax_report_line_cgst','tax_report_line_sgst'))
+                WHERE tag_rep_ln.account_tax_report_line_id in (SELECT res_id FROM ir_model_data WHERE module='l10n_in' AND name in ('tax_report_line_igst','tax_report_line_cgst','tax_report_line_sgst'))
```
Comment (jco-odoo) [inline addons/l10n_in/report/account_invoice_report.py]: Why did you remove the posted?

File: addons/l10n_in/views/account_journal_views.xml [hunk 8]
```diff
@@ -11,6 +11,9 @@
             <xpath expr="//field[@name='loss_account_id']" position="after">
                 <field name="l10n_in_import_export" attrs="{'invisible': [('type', 'not in', ['sale', 'purchase'])]}"/>
             </xpath>
+            <field name="company_id" position="after">
+                <field name="l10n_in_gstin_partner_id" context="{'show_vat':True}" options='{"no_create": True,"always_reload": True}'/>
```
Comment (jco-odoo) [inline addons/l10n_in/views/account_journal_views.xml]: always_reload needed?

File: addons/l10n_in_purchase/models/purchase_order.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        if company_id:
+            company = self.env['res.company'].browse(company_id)
+        else:
+            company = self.env.company
+
+        domain = [('company_id', '=', company.id), ('type', '=', 'purchase')]
+
+        journal = self.env['account.journal'].search(domain, limit=1)
+        if journal:
+            self.l10n_in_journal_id = journal.id
```
Comment (jco-odoo) [inline addons/l10n_in_purchase/models/purchase_order.py]: You have self.company_id and the context is used for default, not onchange.

File: addons/l10n_in_sale/models/sale_order.py [hunk 10]
```diff
@@ -9,8 +9,25 @@ class SaleOrder(models.Model):
...
+        if company_id:
+            company = self.env['res.company'].browse(company_id)
+        else:
+            company = self.env.company
+
+        domain = [('company_id', '=', company.id), ('type', '=', 'sale')]
+
+        journal = self.env['account.journal'].search(domain, limit=1)
+        if journal:
+            self.l10n_in_journal_id = journal.id
```
Comment (jco-odoo) [inline addons/l10n_in_sale/models/sale_order.py]: same as in purchase

File: addons/l10n_in_sale/models/sale_order.py [hunk 11]
```diff
@@ -9,8 +9,25 @@ class SaleOrder(models.Model):
 
     l10n_in_reseller_partner_id = fields.Many2one('res.partner',
         string='Reseller', domain="[('vat', '!=', False), '|', ('company_id', '=', False), ('company_id', '=', company_id)]", states={'posted': [('readonly', True)]})
+    l10n_in_journal_id = fields.Many2one('account.journal', string="Journal", states={'posted': [('readonly', True)]})
 
     def _prepare_invoice(self):
         invoice_vals = super(SaleOrder, self)._prepare_invoice()
-        invoice_vals['l10n_in_reseller_partner_id'] = self.l10n_in_reseller_partner_id.id
```
Comment (jco-odoo) [inline addons/l10n_in_sale/models/sale_order.py]: need to check we don't remove this (nothing to do with multiple operational units)

File: addons/l10n_in/models/res_config_settings.py [hunk 12]
```diff
@@ -10,6 +10,15 @@ class ResConfigSettings(models.TransientModel):
...
 
+    def action_open_company(self):
+        return {
+            'type': 'ir.actions.act_window',
+            'res_model': 'res.company',
+            'view_type': 'form',
+            'view_mode': 'form',
+            'res_id': self.company_id.id,
+        }
+
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/res_config_settings.py]: Now not needed

File: addons/l10n_in/models/res_partner.py (hunk 6 above)
Comment (jva-odoo) [reply addons/l10n_in/models/res_partner.py]: right

File: addons/l10n_in/report/account_invoice_report.py (hunk 7 above)
Comment (jva-odoo) [reply addons/l10n_in/report/account_invoice_report.py]: because now we give filter on the report that all_entery or posted only.

File: addons/l10n_in/views/account_journal_views.xml (hunk 8 above)
Comment (jva-odoo) [reply addons/l10n_in/views/account_journal_views.xml]: because when using show_vat in context then always_reload needed.

**Comment by jva-odoo** [comment]
Transfer to https://github.com/odoo/odoo/pull/37471
